src/Fmt_ast.ml

```diff
@@ -3089,11 +3089,13 @@ and fmt_constructor_declaration c ctx ~max_len_name ~first ~last:_ cstr_decl
          (str pad)
          (fits_breaks ~level:3 "" pad))
   in
-  fmt_if (not first)
-    (match c.conf.type_decl with `Sparse -> "@;<1000 0>" | `Compact -> "@ ")
-  $ Cmts.fmt_before c pcd_loc $ Cmts.fmt_before c loc
+  let has_cmt_before = Cmts.has_before c.cmts pcd_loc in
+  let sparse = Poly.( = ) c.conf.type_decl `Sparse in
+  fmt_if_k (not first) (fmt_or (sparse || has_cmt_before) "@;<1000 0>" "@ ")
+  $ Cmts.fmt_before ~epi:(break 1000 0) c pcd_loc
+  $ Cmts.fmt_before c loc
```

Comment:
It's fine for me but could you add a detailed comment before those changes, and also reference this PR#, in case we need to revert it when we get a more precise comment attachment.

Reply:
Done.